File: internal/builtin/tty.go
package builtin

// node:tty 模块（Phase 2 补充，chalk/supports-color 等包依赖）。
//
//   - isatty(fd) → boolean
//   - ReadStream / WriteStream：非 TTY fd 构造抛 ERR_TTY_INIT_FAILED（Node
//     语义）；原型面 setRawMode / clearLine / cursorTo / getColorDepth 等。
//
// 基于 Go os.File.Stat() 的 ModeCharDevice 判断终端。

import (
	"github.com/aluka-lang/aluka/internal/engine"
)

// NewTTY 构造 node:tty 模块导出。
func NewTTY(ctx engine.Context) (engine.Value, error) {
	m := engine.NewObject()

	_ = m.Set("isatty", engine.NewFunction("isatty", func(args []engine.Value) (engine.Value, error) {
		fd := 0
		if len(args) > 0 {
			if n, ok := args[0].Int(); ok {
				fd = n
			}
		}
		return engine.Boolean(isTTYFd(fd)), nil
	}))

	// ReadStream 原型（setRawMode）。
	rsProto := engine.NewObject()
	_ = rsProto.Set("setRawMode", engine.NewFunction("setRawMode", func(args []engine.Value) (engine.Value, error) {
		return engine.Undefined(), nil
	}))
	// WriteStream 原型（clearLine/clearScreenDown/cursorTo/moveCursor/
	// getColorDepth/hasColors/getWindowSize）。
	wsProto := engine.NewObject()
	mkNoop := func(name string) engine.Value {
		return engine.NewFunction(name, func(args []engine.Value) (engine.Value, error) {
			return engine.Undefined(), nil
		})
	}
	for _, k := range []string{"clearLine", "clearScreenDown", "cursorTo",
		"getColorDepth", "getWindowSize", "hasColors", "moveCursor", "_refreshSize"} {
		_ = wsProto.Set(k, mkNoop(k))
	}
	_ = wsProto.Set("isTTY", engine.Boolean(false))

	// ttyErr 非 TTY 构造错误（Node：ERR_TTY_INIT_FAILED）。
	ttyErr := func() error {
		return &fsCodeError{code: "ERR_TTY_INIT_FAILED",
			msg: "ERR_TTY_INIT_FAILED: TTY initialization failed: uv_tty_init returned EBADF"}
	}

	// ReadStream 构造器。
	rsCtor := engine.NewFunction("ReadStream", func(args []engine.Value) (engine.Value, error) {
		fd := -1
		if len(args) > 0 {
			if n, ok := args[0].Int(); ok {
				fd = n
			}
		}
		if fd >= 0 && !isTTYFd(fd) {
			return engine.Undefined(), ttyErr()
		}
		obj := engine.NewObject()
		isTTY := fd >= 0 && isTTYFd(fd)
		_ = obj.Set("isTTY", engine.Boolean(isTTY))
		_ = obj.Set("fd", engine.IntValue(fd))
		_ = obj.Set("isRaw", engine.Boolean(false))
		_ = obj.Set("setRawMode", engine.NewFunction("setRawMode", func(a []engine.Value) (engine.Value, error) {
			return engine.Undefined(), nil
		}))
		engine.SetProto(obj, rsProto)
		return obj, nil
	})
	if co, ok := rsCtor.AsObject(); ok {
		_ = co.Set("prototype", rsProto)
	}
	_ = rsProto.Set("constructor", rsCtor)

	// WriteStream 构造器。
	wsCtor := engine.NewFunction("WriteStream", func(args []engine.Value) (engine.Value, error) {
		fd := -1
		if len(args) > 0 {
			if n, ok := args[0].Int(); ok {
				fd = n
			}
		}
		if fd >= 0 && !isTTYFd(fd) {
			return engine.Undefined(), ttyErr()
		}
		obj := engine.NewObject()
		isTTY := fd >= 0 && isTTYFd(fd)
		_ = obj.Set("isTTY", engine.Boolean(isTTY))
		_ = obj.Set("fd", engine.IntValue(fd))
		_ = obj.Set("columns", engine.IntValue(80))
		_ = obj.Set("rows", engine.IntValue(24))
		engine.SetProto(obj, wsProto)
		return obj, nil
	})
	if co, ok := wsCtor.AsObject(); ok {
		_ = co.Set("prototype", wsProto)
	}
	_ = wsProto.Set("constructor", wsCtor)

	_ = m.Set("ReadStream", rsCtor)
	_ = m.Set("WriteStream", wsCtor)

	return m, nil
}

// isTTYFd 判断 fd 是否指向终端（平台相关：Windows 用 GetConsoleMode，
// POSIX 用字符设备检测）。
func isTTYFd(fd int) bool {
	return isTTYFdPlatform(fd)
}

File: aluka_r/crates/aluka-vm/src/builtins/tty.rs
//! `tty` 内置模块（Phase 7）：终端交互检测（Node 22 语义）。
//!
//! 语义严格对齐 Go Oracle（`aluka_g/internal/builtin/nodeos/tty.go`）：
//! - `isatty(fd) -> boolean`：fd 0/1/2 经终端检测（Rust 用 std
//!   `IsTerminal`，对齐 Go 的 `ModeCharDevice`/`GetConsoleMode` 判断），
//!   其余 fd 恒为 `false`；非数值 fd 视为 0；
//! - `ReadStream(fd)` / `WriteStream(fd)`：非 TTY 的非负 fd 构造抛
//!   `ERR_TTY_INIT_FAILED`（Node 语义，错误对象带 `code` 字段）；fd 为负
//!   （含缺省 -1）跳过检测直接构造；实例携带 `isTTY`/`fd`（及 ReadStream
//!   的 `isRaw`、WriteStream 的 `columns`/`rows`）；
//! - 原型/实例方法：`setRawMode`（读）与 `clearLine` / `clearScreenDown` /
//!   `cursorTo` / `getColorDepth` / `getWindowSize` / `hasColors` /
//!   `moveCursor` / `_refreshSize`（写）。
//!
//! 已知限制：Rust 的 NativeFn 堆对象不支持挂属性，`ReadStream.prototype` /
//! `WriteStream.prototype` 面以实例自有方法等价提供（Go 亦在实例上重复挂载），
//! `ctor.prototype` 属性未物化，见汇报。

use crate::builtins::{BuiltinRegistry, ModuleDef, register_handler, set_module_prop};
use crate::heap::HeapObject;
use crate::interpreter::{Vm, VmError};
use crate::value::Value;
use aluka_core::ObjectRef;

/// `require("tty")` / `require("node:tty")` 主模块。
pub const MODULE: ModuleDef = ModuleDef { name: "tty", build };

/// 构建 `tty` 模块对象。
fn build(vm: &mut Vm, registry: &mut BuiltinRegistry) -> Result<ObjectRef, VmError> {
    let obj = vm.alloc_ordinary();

    let isatty_fn = vm.alloc_native_fn("tty.isatty");
    set_module_prop(vm, obj, "isatty", Value::Object(isatty_fn))?;
    register_handler(registry, "tty", "isatty", tty_isatty);

    // 构造器（NativeFn 形态注册 `new` 分派；实例方法经 `_builtinNs` 分派）。
    let rs_ctor = vm.alloc_native_fn("tty.ReadStream");
    set_module_prop(vm, obj, "ReadStream", Value::Object(rs_ctor))?;
    register_handler(registry, "tty", "ReadStream", tty_read_stream);
    register_handler(registry, "tty:rs", "setRawMode", tty_set_raw_mode);

    let ws_ctor = vm.alloc_native_fn("tty.WriteStream");
    set_module_prop(vm, obj, "WriteStream", Value::Object(ws_ctor))?;
    register_handler(registry, "tty", "WriteStream", tty_write_stream);
    for method in [
        "clearLine",
        "clearScreenDown",
        "cursorTo",
        "getColorDepth",
        "getWindowSize",
        "hasColors",
        "moveCursor",
        "_refreshSize",
    ] {
        register_handler(registry, "tty:ws", method, tty_noop);
    }

    Ok(obj)
}

/// 实例方法 no-op（返回 `undefined`）。
fn tty_noop(_vm: &mut Vm, _args: &[Value]) -> Result<Value, VmError> {
    Ok(Value::Undefined)
}

/// `rs.setRawMode(...)`：no-op（返回 `undefined`）。
fn tty_set_raw_mode(_vm: &mut Vm, _args: &[Value]) -> Result<Value, VmError> {
    Ok(Value::Undefined)
}

/// `tty.isatty(fd)`：fd 是否指向终端（0/1/2 之外恒 false）。
fn tty_isatty(_vm: &mut Vm, args: &[Value]) -> Result<Value, VmError> {
    // 非数值 fd 视为 0（对齐 Go `Int()` 失败时的默认值）。
    let fd = match args.first() {
        Some(Value::Number(n)) => *n as i64,
        _ => 0,
    };
    Ok(Value::Boolean(is_tty_fd(fd)))
}

/// `new ReadStream(fd)`：TTY 读流构造。
fn tty_read_stream(vm: &mut Vm, args: &[Value]) -> Result<Value, VmError> {
    let fd = arg_fd(args);
    if fd >= 0 && !is_tty_fd(fd) {
        return Err(tty_init_failed(vm));
    }
    let obj = vm.alloc_ordinary();
    let ns = Value::Object(vm.alloc_string("tty:rs".to_owned()));
    set_module_prop(vm, obj, "_builtinNs", ns)?;
    set_module_prop(vm, obj, "isTTY", Value::Boolean(fd >= 0 && is_tty_fd(fd)))?;
    set_module_prop(vm, obj, "fd", Value::Number(fd as f64))?;
    set_module_prop(vm, obj, "isRaw", Value::Boolean(false))?;
    let set_raw = vm.alloc_native_fn("tty:rs.setRawMode");
    set_module_prop(vm, obj, "setRawMode", Value::Object(set_raw))?;
    Ok(Value::Object(obj))
}

/// `new WriteStream(fd)`：TTY 写流构造。
fn tty_write_stream(vm: &mut Vm, args: &[Value]) -> Result<Value, VmError> {
    let fd = arg_fd(args);
    if fd >= 0 && !is_tty_fd(fd) {
        return Err(tty_init_failed(vm));
    }
    let obj = vm.alloc_ordinary();
    let ns = Value::Object(vm.alloc_string("tty:ws".to_owned()));
    set_module_prop(vm, obj, "_builtinNs", ns)?;
    set_module_prop(vm, obj, "isTTY", Value::Boolean(fd >= 0 && is_tty_fd(fd)))?;
    set_module_prop(vm, obj, "fd", Value::Number(fd as f64))?;
    set_module_prop(vm, obj, "columns", Value::Number(80.0))?;
    set_module_prop(vm, obj, "rows", Value::Number(24.0))?;
    for method in [
        "clearLine",
        "clearScreenDown",
        "cursorTo",
        "getColorDepth",
        "getWindowSize",
        "hasColors",
        "moveCursor",
        "_refreshSize",
    ] {
        let fn_ref = vm.alloc_native_fn(&format!("tty:ws.{method}"));
        set_module_prop(vm, obj, method, Value::Object(fn_ref))?;
    }
    Ok(Value::Object(obj))
}

/// 提取 fd 实参（缺省 -1，非数值 -1——对齐 Go `Int()` 失败时的默认）。
fn arg_fd(args: &[Value]) -> i64 {
    match args.first() {
        Some(Value::Number(n)) => *n as i64,
        _ => -1,
    }
}

/// 非 TTY fd 构造错误（Node：`ERR_TTY_INIT_FAILED`）。
fn tty_init_failed(vm: &mut Vm) -> VmError {
    let obj = vm.alloc_ordinary();
    let name_v = Value::Object(vm.alloc_string("Error".to_owned()));
    let _ = vm.set_property(Value::Object(obj), "name", name_v);
    let msg_v = Value::Object(vm.alloc_string(
        "ERR_TTY_INIT_FAILED: TTY initialization failed: uv_tty_init returned EBADF".to_owned(),
    ));
    let _ = vm.set_property(Value::Object(obj), "message", msg_v);
    let code_v = Value::Object(vm.alloc_string("ERR_TTY_INIT_FAILED".to_owned()));
    let _ = vm.set_property(Value::Object(obj), "code", code_v);
    VmError::Thrown(Value::Object(obj))
}

/// 平台终端检测：0/1/2 经 std `IsTerminal`（Windows 走 GetConsoleMode 等价
/// 判定），其余 fd 恒 false。
fn is_tty_fd(fd: i64) -> bool {
    use std::io::IsTerminal;
    match fd {
        0 => std::io::stdin().is_terminal(),
        1 => std::io::stdout().is_terminal(),
        2 => std::io::stderr().is_terminal(),
        _ => false,
    }
}

/// 判断堆对象形态的辅助（保持对 `HeapObject` 的窄依赖）。
#[allow(dead_code)]
fn is_heap_object(vm: &Vm, r: ObjectRef) -> bool {
    vm.heap
        .get(r.0 as usize)
        .is_some_and(|o| matches!(o, HeapObject::Ordinary { .. }))
}

/// 编译期锚定：处理器签名与注册表一致。
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn handler_signatures_anchor() {
        let _: crate::builtins::BuiltinHandler = tty_isatty;
        let _: crate::builtins::BuiltinHandler = tty_read_stream;
        let _: crate::builtins::BuiltinHandler = tty_write_stream;
    }

    #[test]
    fn isatty_negative_fd_is_false() {
        assert!(!is_tty_fd(-1));
        assert!(!is_tty_fd(99));
    }
}
